components/brave_wallet/browser/solana_tx_manager.cc

```diff
@@ -631,6 +640,191 @@ void SolanaTxManager::MakeTxDataFromBase64EncodedTransaction(
+    MakeBubbleGumProgramTransferTxDataCallback callback,
+    std::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess) {
+    std::move(callback).Run(nullptr, error, error_message);
+    return;
+  }
+
+  auto account_data_bytes = base::Base64Decode(account_info->data);
+  if (!account_data_bytes) {
+    std::move(callback).Run(
+        nullptr, mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto result = DecodeMerkleTreeAuthorityAndDepth(*account_data_bytes);
+  if (!result) {
+    std::move(callback).Run(
+        nullptr, mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  std::optional<SolanaInstruction> instruction =
+      solana::bubblegum_program::Transfer(
+          result->first, result->second.ToBase58(), to_wallet_address, proof);
+
+  if (!instruction) {
+    std::move(callback).Run(
+        nullptr, mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  std::vector<SolanaInstruction> vec;
+  vec.emplace_back(std::move(instruction.value()));
+
+  // recent_blockhash will be updated when we are going to send out the tx.
+  auto msg = SolanaMessage::CreateLegacyMessage("" /* recent_blockhash*/, 0,
+                                                // from_wallet_address,
+                                                proof.owner,  // Should probably
+                                                std::move(vec));
+  if (!msg) {
+    std::move(callback).Run(
+        nullptr, mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  SolanaTransaction transaction(std::move(*msg));
+  transaction.set_to_wallet_address(to_wallet_address);
+  transaction.set_tx_type(mojom::TransactionType::SolanaCompressedNftTransfer);
+  // transaction.set_lamports(lamports); // why not ?
+  auto tx_data = transaction.ToSolanaTxData();
+  DCHECK(tx_data);
+  std::move(callback).Run(std::move(tx_data),
+                          mojom::SolanaProviderError::kSuccess, "");
+}
+
+// Adapted from
+// https://github.com/solana-labs/solana-program-library/blob/master/account-compression/sdk/src/accounts/ConcurrentMerkleTreeAccount.ts#L140
+std::optional<std::pair<uint32_t, SolanaAddress>>
+SolanaTxManager::DecodeMerkleTreeAuthorityAndDepth(
+    const std::vector<uint8_t>& data) {
+  size_t offset = 0;
+
+  // Decode the first byte, representing the compression account type .
+  // The possible values are 0=Unitialized and 1=ConcurrentMerkleTree.
+  // The value must be 1.
+  auto compression_account_type =
+      solana_ins_data_decoder::DecodeUint8(data, offset);
+  if (!compression_account_type) {
+    return std::nullopt;
+  }
```

Comment:
Let's keep it that way, but there is now base::SpanReader which would be a preferrable choice.